File: python/sketchatone/tablet/server/event_adapter.py
"""Convert OTD report objects into the websocket event shape the visualizer consumes."""

from __future__ import annotations

import math
import time
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Tuple

from ..otd.config_loader import TabletConfiguration
from ..otd.reports import (
    DeviceReport,
    OutOfRangeReport,
    TabletReport,
    TiltTabletReport,
)


@dataclass
class TabletEvent:
    """Normalized tablet event broadcast to clients.

    Coordinate ranges:
        x, y       0.0 .. 1.0 (origin top-left)
        pressure   0.0 .. 1.0
        tiltX, Y   -1.0 .. 1.0
        tiltXY     0.0 .. 1.0 (magnitude)
    """
    type: str = "tablet-data"
    timestamp: float = 0.0
    state: str = "none"
    x: float = 0.0
    y: float = 0.0
    pressure: float = 0.0
    tiltX: float = 0.0
    tiltY: float = 0.0
    tiltXY: float = 0.0
    primaryButtonPressed: bool = False
    secondaryButtonPressed: bool = False
    auxCodes: List[int] = field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "type": self.type,
            "timestamp": self.timestamp,
            "state": self.state,
            "x": self.x,
            "y": self.y,
            "pressure": self.pressure,
            "tiltX": self.tiltX,
            "tiltY": self.tiltY,
            "tiltXY": self.tiltXY,
            "primaryButtonPressed": self.primaryButtonPressed,
            "secondaryButtonPressed": self.secondaryButtonPressed,
            "auxCodes": list(self.auxCodes),
        }


class EventAdapter:
    """Stateful adapter: combines successive tablet + aux reports into events."""

    def __init__(self, config: TabletConfiguration):
        self.config = config
        self._max_x = max(1, config.specifications.digitizer_max_x)
        self._max_y = max(1, config.specifications.digitizer_max_y)
        self._max_p = max(1, config.specifications.pen_max_pressure)
        self._aux_codes_by_slot: Dict[Tuple[bytes, int], List[int]] = {}
        self._in_range: bool = False

    def override_ranges(self,
                        max_x: Optional[int] = None,
                        max_y: Optional[int] = None,
                        max_pressure: Optional[int] = None) -> None:
        """Replace the normalization denominators (e.g. with values read from
        the device's HID report descriptor when the OTD config's vendor-mode
        values don't match the interface we actually opened)."""
        if max_x:
            self._max_x = max(1, int(max_x))
        if max_y:
            self._max_y = max(1, int(max_y))
        if max_pressure:
            self._max_p = max(1, int(max_pressure))

    @staticmethod
    def _now_ms() -> float:
        return time.time() * 1000.0

    def _current_aux_codes(self) -> List[int]:
        seen: set = set()
        out: List[int] = []
        for codes in self._aux_codes_by_slot.values():
            for c in codes:
                if c not in seen:
                    seen.add(c)
                    out.append(c)
        return out

    def update_aux_codes(self, path: bytes, report_id: int, codes: List[int]) -> bool:
        """Record the active scan codes for one (path, report_id) slot.

        Composite aux interfaces multiplex several top-level collections
        over a single IOHIDDevice; keying state per report ID prevents a
        release on one collection from clobbering held buttons on
        another. Returns True if the union across all slots changed.
        """
        prev = self._current_aux_codes()
        slot = (path, report_id)
        if codes:
            self._aux_codes_by_slot[slot] = list(codes)
        else:
            self._aux_codes_by_slot.pop(slot, None)
        return self._current_aux_codes() != prev

    def _empty_event(self) -> TabletEvent:
        ev = TabletEvent(timestamp=self._now_ms())
        ev.auxCodes = self._current_aux_codes()
        return ev

    def adapt(self, report: Any) -> Optional[TabletEvent]:
        """Translate a parsed OTD report into a TabletEvent, or None to skip."""
        if isinstance(report, OutOfRangeReport):
            self._in_range = False
            return self._empty_event()

        if isinstance(report, (TabletReport, TiltTabletReport)) or hasattr(report, "position"):
            self._in_range = True
            x_raw, y_raw = report.position
            pressure_raw = getattr(report, "pressure", 0)
            ev = self._empty_event()
            ev.x = min(1.0, max(0.0, x_raw / self._max_x))
            ev.y = min(1.0, max(0.0, y_raw / self._max_y))
            ev.pressure = min(1.0, max(0.0, pressure_raw / self._max_p))
            pen_buttons = list(getattr(report, "pen_buttons", []) or [])
            ev.primaryButtonPressed = len(pen_buttons) > 0 and pen_buttons[0]
            ev.secondaryButtonPressed = len(pen_buttons) > 1 and pen_buttons[1]
            if hasattr(report, "tilt"):
                tx, ty = report.tilt
                ev.tiltX = max(-1.0, min(1.0, tx / 127.0))
                ev.tiltY = max(-1.0, min(1.0, ty / 127.0))
                ev.tiltXY = min(1.0, math.hypot(ev.tiltX, ev.tiltY))
            ev.state = "contact" if ev.pressure > 0 else "hover"
            return ev

        if isinstance(report, DeviceReport):
            return None

        return None

File: src/tablet/server/eventAdapter.ts
/** Convert OTD report objects into the websocket event shape the visualizer consumes. */

import { TabletConfiguration } from "../otd/configLoader.js";
import { ParsedReport } from "../otd/parsers/index.js";

/**
 * Normalized tablet event broadcast to clients.
 *
 * Coordinate ranges:
 *     x, y       0.0 .. 1.0 (origin top-left)
 *     pressure   0.0 .. 1.0
 *     tiltX, Y   -1.0 .. 1.0
 *     tiltXY     0.0 .. 1.0 (magnitude)
 */
export interface TabletEvent {
  type: "tablet-data";
  timestamp: number;
  state: "none" | "hover" | "contact";
  x: number;
  y: number;
  pressure: number;
  tiltX: number;
  tiltY: number;
  tiltXY: number;
  primaryButtonPressed: boolean;
  secondaryButtonPressed: boolean;
  auxCodes: number[];
}

function blankEvent(): TabletEvent {
  return {
    type: "tablet-data",
    timestamp: 0,
    state: "none",
    x: 0, y: 0, pressure: 0,
    tiltX: 0, tiltY: 0, tiltXY: 0,
    primaryButtonPressed: false,
    secondaryButtonPressed: false,
    auxCodes: [],
  };
}

function clamp01(n: number): number { return Math.min(1.0, Math.max(0.0, n)); }
function clampUnit(n: number): number { return Math.max(-1.0, Math.min(1.0, n)); }

function hasPosition(r: ParsedReport): r is ParsedReport & {
  position: [number, number];
  pressure?: number;
  penButtons?: boolean[];
  tilt?: [number, number];
} {
  return (r as any).position !== undefined;
}

/** Stateful adapter: combines successive tablet + aux reports into events. */
export class EventAdapter {
  private maxX: number;
  private maxY: number;
  private maxP: number;
  private auxCodesBySlot = new Map<string, number[]>();
  private inRange = false;

  constructor(public config: TabletConfiguration) {
    this.maxX = Math.max(1, config.specifications.digitizerMaxX);
    this.maxY = Math.max(1, config.specifications.digitizerMaxY);
    this.maxP = Math.max(1, config.specifications.penMaxPressure);
  }

  /**
   * Replace the normalization denominators (e.g. with values read from
   * the device's HID report descriptor when the OTD config's vendor-mode
   * values don't match the interface we actually opened).
   */
  overrideRanges(opts: { maxX?: number; maxY?: number; maxPressure?: number }): void {
    if (opts.maxX) this.maxX = Math.max(1, Math.trunc(opts.maxX));
    if (opts.maxY) this.maxY = Math.max(1, Math.trunc(opts.maxY));
    if (opts.maxPressure) this.maxP = Math.max(1, Math.trunc(opts.maxPressure));
  }

  private nowMs(): number { return Date.now(); }

  currentAuxCodes(): number[] {
    const seen = new Set<number>();
    const out: number[] = [];
    for (const codes of this.auxCodesBySlot.values()) {
      for (const c of codes) {
        if (!seen.has(c)) { seen.add(c); out.push(c); }
      }
    }
    return out;
  }

  /**
   * Record the active scan codes for one (path, reportId) slot.
   *
   * Composite aux interfaces multiplex several top-level collections
   * over a single IOHIDDevice; keying state per report ID prevents a
   * release on one collection from clobbering held buttons on
   * another. Returns true if the union across all slots changed.
   */
  updateAuxCodes(path: string, reportId: number, codes: number[]): boolean {
    const prev = this.currentAuxCodes();
    const slot = `${path}:${reportId}`;
    if (codes.length > 0) {
      this.auxCodesBySlot.set(slot, [...codes]);
    } else {
      this.auxCodesBySlot.delete(slot);
    }
    const next = this.currentAuxCodes();
    if (prev.length !== next.length) return true;
    for (let i = 0; i < prev.length; i++) if (prev[i] !== next[i]) return true;
    return false;
  }

  emptyEvent(): TabletEvent {
    const ev = blankEvent();
    ev.timestamp = this.nowMs();
    ev.auxCodes = this.currentAuxCodes();
    return ev;
  }

  /** Translate a parsed OTD report into a TabletEvent, or null to skip. */
  adapt(report: ParsedReport): TabletEvent | null {
    if (report.kind === "OutOfRangeReport") {
      this.inRange = false;
      return this.emptyEvent();
    }
    if (hasPosition(report)) {
      this.inRange = true;
      const [xRaw, yRaw] = report.position;
      const ev = this.emptyEvent();
      ev.x = clamp01(xRaw / this.maxX);
      ev.y = clamp01(yRaw / this.maxY);
      ev.pressure = clamp01((report.pressure ?? 0) / this.maxP);
      const penButtons = report.penButtons ?? [];
      ev.primaryButtonPressed = penButtons.length > 0 && !!penButtons[0];
      ev.secondaryButtonPressed = penButtons.length > 1 && !!penButtons[1];
      if (report.tilt) {
        const [tx, ty] = report.tilt;
        ev.tiltX = clampUnit(tx / 127.0);
        ev.tiltY = clampUnit(ty / 127.0);
        ev.tiltXY = Math.min(1.0, Math.hypot(ev.tiltX, ev.tiltY));
      }
      ev.state = ev.pressure > 0 ? "contact" : "hover";
      return ev;
    }
    return null;
  }
}
